dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/GitHubSourceCodeRepo.java

```diff
@@ -1127,55 +1138,104 @@ public String getDefaultBranch(String repositoryId) {
-                for (GHCommit ghCommit : ghCommits) {
-                    try {
-                        // this will only resolve if a .dockstore.yml was in the last commit to be touched (is not eligible for cache since it uses JWT)
-                        ghCommit.listBranchesWhereHead().toList().forEach(branch -> candidateBranches.add("refs/heads/" + branch.getName()));
-                    } catch (IOException e) {
-                        // do nothing and proceed to next commit
-                        LOG.info("had issue processing branch for .dockstore.yml", e);
-                    }
-                }
-                // examine the default branch, good chance it will have it
-                String defaultBranch = "refs/heads/" + getDefaultBranch(repositoryId);
-                examineBranchForDockstoreYml(repositoryId, candidateBranches, defaultBranch);
-                // if we still don't have a candidate try guessing at some arbitrary (but a constant number of) branches. Hopefully this does not trigger too often
-                if (candidateBranches.isEmpty()) {
-                    String[] totalGuesses = {"master", "main", "develop"};
-                    Arrays.stream(totalGuesses).toList().stream().map(guess -> "refs/heads/" + guess).forEach(guess -> {
-                        examineBranchForDockstoreYml(repositoryId, candidateBranches, guess);
-                    });
-                }
-            } catch (IOException e) {
-                LOG.error("Unable to retrieve analyze branch candidates for repository " + repositoryId, e);
-                return candidateBranches;
+        final int maxResults = 5;
+        final Set<String> likelies = Set.of("master", "main", "develop");
+
+        try {
+            if (repositoryId == null) {
+                return Set.of();
+            }
+            // Get repository and default branch.
+            GHRepository repository = getRepository(repositoryId);
+            String defaultBranch = repository.getDefaultBranch();
+
+            // Get the branch names, ordered so that branches with recent pushes are first.
+            List<String> branches = getBranchesWithRecentPushesFirst(repository);
+
+            // Reorder the branch names so the default branch comes first, followed by any "likely" branches, then the rest.
+            // Do so by first moving the likelies to the front, and then moving the default branch to the front.
+            branches = listTrueFirst(branches, likelies::contains);
+            branches = listTrueFirst(branches, branch -> Objects.equals(defaultBranch, branch));
+
+            // For a subset of branches at the front of the list, return a set of the corresponding refs that contain a .dockstore.yml.
+            return branches.stream()
+                .limit(maxResults)
+                .map(branch -> REFS_HEADS + branch)
```

Comment:
I'm probably missing something, but it seems like you strip out `REF_HEADS` on line 1202 only to add it back here. Is there a reason for that?

Reply:
"Historical reasons" :)

Originally, I used a different method to retrieve the branch names (`getBranches`), and standardized that section of code to use branch names rather than refs.  Later, I changed the code to use `listRefs`, because `getBranches` gets _all_ the branches, which could consume many API calls, whereas `listRefs` returns an iterator, so we can limit the number.  Turns out that `getDefaultBranch` also returns a branch name, so I left the code the way it was...